text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+the computed property is, but noticing missing dependencies is a challenge even
+for experienced Ember developers when the computed property gets more
+complicated.
+
+Tracked properties have a feature called _autotrack_, where dependencies are
+automatically detected as they are used. This means that as long as all
+dependencies are marked as tracked, they will automatically be detected:
+
+```js
+import { tracked } from '@glimmer/tracking';
+
+class Person {
+  @tracked firstName = 'Tom';
+  @tracked lastName = 'Dale';
+
+  @tracked
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+This also allows us to opt out of tracking entirely, like if we know for
+instance that a given property is constant and will never change. In general,
+the idea is that _mutable_ properties should be marked as tracked, and
+_immutable_ properties should not.
+
+### Reducing Memory Consumption
+
+By default, computed properties cache their values. This is great when a
+computed property has to perform expensive work to produce its value, and that
+value gets used over and over again.
+
+But checking, populating, and invalidating this cache comes with its own
+overhead. Modern JavaScript VMs can produce highly optimized code, and in many
+cases the overhead of caching is greater than the cost of simply recomputing the
+value.
+
+Worse, cached computed property values cannot be freed by the garbage collector
+until the entire object is freed. Many computed properties are accessed only
+once, but because they cache by default, they take up valuable space on the heap
+for no benefit.
+
+For example, imagine this component that checks whether the `files` property is
+supported in input elements:
+
+```js
+import Component from '@ember/component';
+import { computed } from '@ember/object';
+
+export default Component.extend({
+  inputElement: computed(function() {
+    return document.createElement('input');
+  }),
+
+  supportsFiles: computed('inputElement', function() {
+    return 'files' in this.inputElement;
+  }),
+
+  didInsertElement() {
+    if (this.supportsFiles) {
+      // do something
+    } else {
+      // do something else
+    }
+  },
+});
+```
+
+This component would create and retain an `HTMLInputElement` DOM node for the
+lifetime of the component, even though all we really want to cache is the
+Boolean value of whether the browser supports the `files` attribute.
+
+Particularly on mobile devices, where RAM is limited and often slow, we should
+be more conservative about our memory consumption. Tracked properties switch
+from an opt-out caching model to opt-in, allowing developers to err on the side
+of reduced memory usage, but easily enabling caching (a.k.a. memoization) if a
+property shows up as a bottleneck during profiling.
+
+## Detailed Design
+
+This RFC proposes adding the `tracked` decorator function, used to mark
+properties and getters as tracked:
+
+```ts
+const tracked: PropertyDecorator;
```

Comment:
what's the definition of `PropertyDecorator`?  is it a new built in that's coming with stage 3?
or:
```ts
type PropertyDecorator = (property: ElementDescripiptor) => ElementDescriptor
```
?

Reply:
There is a [`PropertyDecorator` in `lib.es5.d.ts`](https://github.com/Microsoft/TypeScript/blob/eb2297df022f6b1f284aff96e93c06097bb01ea5/lib/lib.es5.d.ts#L1370), but the type is using the outdated TypeScript signature.

The stage 2 decorators only receive single `ElementDescriptor` and can return an enhanced version of it, like you correctly showed in you code sample.

All properties are documented here: https://github.com/tc39/proposal-decorators/blob/master/TABLE.md